### DISCUSSION TITLE
Replace XHR requests with `fetch

### DISCUSSION THREAD

**Discussion by smellyshovel** [maplibre/maplibre-gl-js#2004] [Ideas]
Found this code in `src/util/ajax.ts`.

```
export const makeRequest = function(requestParameters: RequestParameters, callback: ResponseCallback<any>): Cancelable {
    // We're trying to use the Fetch API if possible. However, in some situations we can't use it:
    // - IE11 doesn't support it at all. In this case, we dispatch the request to the main thread so
    //   that we can get an accruate referrer header.
    // - Safari exposes window.AbortController, but it doesn't work actually abort any requests in
    //   some versions (see https://bugs.webkit.org/show_bug.cgi?id=174980#c2)
    // - Requests for resources with the file:// URI scheme don't work with the Fetch API either. In
    //   this case we unconditionally use XHR on the current thread since referrers don't matter.
    if (/:\/\//.test(requestParameters.url) && !(/^https?:|^file:/.test(requestParameters.url))) {
        if (isWorker() && (self as any).worker && (self as any).worker.actor) {
            return (self as any).worker.actor.send('getResource', requestParameters, callback);
        }
        if (!isWorker()) {
            const protocol = requestParameters.url.substring(0, requestParameters.url.indexOf('://'));
            const action = config.REGISTERED_PROTOCOLS[protocol] || makeFetchRequest;
            return action(requestParameters, callback);
        }
    }
    if (!isFileURL(requestParameters.url)) {
        if (fetch && Request && AbortController && Object.prototype.hasOwnProperty.call(Request.prototype, 'signal')) {
            return makeFetchRequest(requestParameters, callback);
        }
        if (isWorker() && (self as any).worker && (self as any).worker.actor) {
            const queueOnMainThread = true;
            return (self as any).worker.actor.send('getResource', requestParameters, callback, undefined, queueOnMainThread);
        }
    }
    return makeXMLHttpRequest(requestParameters, callback);
};
```

The comments are fairly interesting in that that I feel like they are not quite relevant anymore.

- IE11 doesn't support it at all. In this case, we dispatch the request to the main thread so that we can get an accruate referrer header.
  **Is it really still relevant?**
- Safari exposes window.AbortController, but it doesn't work actually abort any requests in some versions (see https://bugs.webkit.org/show_bug.cgi?id=174980#c2)
  **The bug was fixed back in 2019**
- Requests for resources with the file:// URI scheme don't work with the Fetch API either. In this case we unconditionally use XHR on the current thread since referrers don't matter.
  **Where does the map send requests with `file://` protocol? Only in dev. mode I believe? Can't they be proxied instead?** 

The main desire to remove XHR requests alltogether is conditioned by the desire to use `fetch` API instead, which uses Promises natively under the hood. And there were ideas to rewrite some major callback-related parts of the code to use Promises instead.

So are we OK with moving from XHR to fetch, @HarelM? Not that it's really crucial, but I see it a good point to refactor.

**Comment by HarelM** [comment]
You have some good points here, IE11 is no longer supported, and the safari might be solved, but in cordova/ionic environment file:// is used so I think xhr is needed (for example loading a style that is on the device).
Having said that, there's no problem in wrapping the xhr code in a promise.
The problem is that I'm not aware of ways to cancel a promise, so I'm not sure xhr would be my first place to try and fix.
In general, there are many places that use callbacks that their code can be simplified a lot using promises.

**Comment by smellyshovel** [comment]
> The problem is that I'm not aware of ways to cancel a promise

I thought more about reurning an object where the request result is wrapped into a Promise, and separately a cancel method.

> but in cordova/ionic environment file:// is used so I think xhr is needed (for example loading a style that is on the device).

OK, makes sense. Thanks.
